server/channels/api4/remote_cluster.go

```diff
@@ -293,3 +301,359 @@ func remoteSetProfileImage(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	ReturnStatusOK(w)
 }
+
+func getRemoteClusters(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionManageSecureConnections) {
+		c.SetPermissionError(model.PermissionManageSecureConnections)
+		return
+	}
+
+	// make sure remote cluster service is enabled.
+	if _, appErr := c.App.GetRemoteClusterService(); appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	filter := model.RemoteClusterQueryFilter{
+		ExcludeOffline: c.Params.ExcludeOffline,
+		InChannel:      c.Params.InChannel,
+		NotInChannel:   c.Params.NotInChannel,
+		Topic:          c.Params.Topic,
+		CreatorId:      c.Params.CreatorId,
+		OnlyConfirmed:  c.Params.OnlyConfirmed,
+		PluginID:       c.Params.PluginId,
+		OnlyPlugins:    c.Params.OnlyPlugins,
+		ExcludePlugins: c.Params.ExcludePlugins,
+	}
+
+	rcs, appErr := c.App.GetAllRemoteClusters(c.Params.Page, c.Params.PerPage, filter)
+	if appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	for _, rc := range rcs {
+		rc.Sanitize()
+	}
+
+	b, err := json.Marshal(rcs)
+	if err != nil {
+		c.Err = model.NewAppError("getRemoteClusters", "api.marshal_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		return
+	}
+
+	w.Write(b)
+}
+
+func createRemoteCluster(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionManageSecureConnections) {
+		c.SetPermissionError(model.PermissionManageSecureConnections)
+		return
+	}
+
+	// make sure remote cluster service is enabled.
+	if _, appErr := c.App.GetRemoteClusterService(); appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	auditRec := c.MakeAuditRecord("createRemoteCluster", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+
+	var rcWithTeamAndPassword model.RemoteClusterWithPassword
+	if jsonErr := json.NewDecoder(r.Body).Decode(&rcWithTeamAndPassword); jsonErr != nil {
+		c.SetInvalidParamWithErr("remoteCluster", jsonErr)
+		return
+	}
+
+	if rcWithTeamAndPassword.Password == "" {
+		c.SetInvalidParam("password")
+		return
+	}
+
+	url := c.App.GetSiteURL()
+	if url == "" {
+		c.Err = model.NewAppError("createRemoteCluster", "api.get_site_url_error", nil, "", http.StatusUnprocessableEntity)
+		return
+	}
+
+	if rcWithTeamAndPassword.DisplayName == "" {
+		rcWithTeamAndPassword.DisplayName = rcWithTeamAndPassword.Name
+	}
+
+	rc := &model.RemoteCluster{
+		Name:        rcWithTeamAndPassword.Name,
+		DisplayName: rcWithTeamAndPassword.DisplayName,
+		SiteURL:     model.SiteURLPending + model.NewId(),
+		Token:       model.NewId(),
+		CreatorId:   c.AppContext.Session().UserId,
+	}
+
+	audit.AddEventParameterAuditable(auditRec, "remotecluster", rc)
+
+	rcSaved, appErr := c.App.AddRemoteCluster(rc)
+	if appErr != nil {
+		c.Err = appErr
+		return
+	}
+	rcSaved.Sanitize()
+
+	inviteCode, iErr := c.App.CreateRemoteClusterInvite(rcSaved.RemoteId, url, rcSaved.Token, rcWithTeamAndPassword.Password)
```

Comment:
Should we move this out of the app layer and put it in the service?  Can be something we do after this PR.  I've been thinking that having a bunch of shared channels APIs in the app layer is just plumbing overhead.

Reply:
Agree, and there are some asymmetries here, as some endpoints on the API layer / slash commands directly call the service, and others go through the app layer. I've created https://mattermost.atlassian.net/browse/MM-59048 to track this